test/val/val_tensor_test.cpp

```diff
@@ -0,0 +1,1195 @@
+            %float = OpTypeFloat 32
+             %fnty = OpTypeFunction %void
+        %uint_vec4 = OpTypeVector %uint 4
+           %uint_0 = OpConstant %uint 0
+      %uint_0_spec = OpSpecConstant %uint 0
+           %uint_1 = OpConstant %uint 1
+           %uint_2 = OpConstant %uint 2
+           %uint_3 = OpConstant %uint 3
+           %uint_4 = OpConstant %uint 4
+          %uint_42 = OpConstant %uint 42
+       %uint_1_bis = OpConstant %uint 1
+          %float_1 = OpConstant %float 1
+%uint_vec4_1_1_1_1 = OpConstantComposite %uint_vec4 %uint_1 %uint_1 %uint_1 %uint_1
+        %uint_arr1 = OpTypeArray %uint %uint_1
+        %uint_arr2 = OpTypeArray %uint %uint_2
+        %uint_arr3 = OpTypeArray %uint %uint_3
+        %uint_arr4 = OpTypeArray %uint %uint_4
+       %float_arr4 = OpTypeArray %float %uint_4
+%uint_arr4_1_1_1_1 = OpConstantComposite %uint_arr4 %uint_1 %uint_1 %uint_1 %uint_1
+%uint_arr4_0_0_0_0_spec = OpSpecConstantComposite %uint_arr4 %uint_0_spec %uint_0_spec %uint_0_spec %uint_0_spec
+    %uint_arr2_1_1 = OpConstantComposite %uint_arr2 %uint_1 %uint_1
+%float_arr4_1_1_1_1 = OpConstantComposite %float_arr4 %float_1 %float_1 %float_1 %float_1
+%uint_arr4_1_1_0_1 = OpConstantComposite %uint_arr4 %uint_1 %uint_1 %uint_0 %uint_1
+ %uint_ptr_Private = OpTypePointer Private %uint
+%uint_arr4_ptr_Private = OpTypePointer Private %uint_arr4
+       %uint_var_1 = OpVariable %uint_ptr_Private Private %uint_1
+%var_uint_arr4_1_1_1_1 = OpVariable %uint_arr4_ptr_Private Private %uint_arr4_1_1_1_1
+    %tensor_uint_4 = OpTypeTensorARM %uint %uint_4
+     %tensor_float = OpTypeTensorARM %float
+%tensor_uint_4_ptr_UniformConstant = OpTypePointer UniformConstant %tensor_uint_4
+%tensor_var = OpVariable %tensor_uint_4_ptr_UniformConstant UniformConstant
+%tensor_float_ptr_UniformConstant = OpTypePointer UniformConstant %tensor_float
+%tensor_var_float_unranked = OpVariable %tensor_float_ptr_UniformConstant UniformConstant
+)";
+  const std::string footer = R"(
+             %fnep = OpFunction %void None %fnty
+            %label = OpLabel
+                     OpReturn
+                     OpFunctionEnd
+)";
+  return header + body + footer;
+}
+
+//
+// Type tests
+//
+
+TEST_F(ValidateTensor, ValidTypeElementTypeOnly) {
+  const std::string src = R"(
+    %test_type = OpTypeTensorARM %uint
+)";
+  std::string spvasm = GenerateModule(src);
+  CompileSuccessfully(spvasm, SPVENV);
+  EXPECT_EQ(SPV_SUCCESS, ValidateInstructions(SPVENV));
+}
+
+TEST_F(ValidateTensor, InvalidTypeTensorElementTypeNotScalar) {
+  const std::string src = R"(
+    %test_type = OpTypeTensorARM %void
+)";
+  std::string spvasm = GenerateModule(src);
+  CompileSuccessfully(spvasm, SPVENV);
+  EXPECT_EQ(SPV_ERROR_INVALID_ID, ValidateInstructions(SPVENV));
+  EXPECT_THAT(getDiagnosticString(),
+              MatchesRegex("OpTypeTensorARM Element Type <id> "
+                           "'[[:digit:]]\\[%void\\]' is not a scalar type.*"));
```

Comment:
It seems this syntax tripped up Effcee.
Use `\d` or `[0-9]` instead of `[[:digit:]]`.  I've seen both those in existing tests.

Reply:
I've updated the PR to use `[0-9]`. It seems this isn't accepted on Windows either. I'm seeing the following in logs:
```
Syntax error at index 133 in simple regular expression "OpConstantComposite Constituent <id> '[0-9]+\[%[0-9]+\]' must have a Rank that is 1 less than the Rank of Result Type <id> '[0-9]+\[%[0-9]+\]', expected 1 but got 2.*": '[' is unsupported
```
I can't find any other uses of `[0-9]` or `\d` in validator tests. I'll try `\d`.

Reply:
It seems that Googletest doesn't support the same regular expression syntax on all platforms (see http://google.github.io/googletest/advanced.html#regular-expression-syntax). Unless I'm misunderstanding, their recommendation is to test for a preprocessor macro that they provide to select the syntax in tests. Not great...

Reply:
FYI.  In the extreme you can use effcee in tests, like the optimizer does.
effcee is my library clone of core LLVM FileCheck functionality.